src/target/lpc55xx.c

```diff
@@ -37,12 +37,515 @@
+	}
+}
+
+static target_addr_t lpc55xx_get_ffr_get_uuid_address(target_s *target)
+{
+	target_addr_t bootloader_tree_address = lpc55xx_get_bootloader_tree_address(target);
+	target_addr_t flash_table_address = lpc55xx_get_flash_table_address(target, bootloader_tree_address);
+
+	if (lpc55xx_get_rom_api_version(target, bootloader_tree_address) == 0) {
+		return target_mem_read32(target, flash_table_address + 10 * sizeof(uint32_t));
+	} else {
+		return target_mem_read32(target, flash_table_address + 13 * sizeof(uint32_t));
+	}
+}
+
+static enum iap_status iap_call_raw(target_s *target, enum iap_cmd cmd, uint32_t r1, uint32_t r2, uint32_t r3)
+{
+	/* Prepare the registers for the IAP call. R0 is always flash_config */
+
+	uint32_t regs[target->regs_size / sizeof(uint32_t)];
+	target_regs_read(target, regs);
+
+	regs[REG_MSP] = LPC55xx_IAP_MSP_ADDRESS;
+	regs[0] = LPC55xx_FLASH_CONFIG_ADDRESS;
+	regs[1] = r1;
+	regs[2] = r2;
+	regs[3] = r3;
+
+	/* Locate the correct IAP function address based on silicon revision */
+
+	switch (cmd) {
+	case IAP_CMD_FLASH_INIT:
+		regs[REG_PC] = lpc55xx_get_flash_init_address(target);
+		break;
+	case IAP_CMD_FLASH_ERASE:
+		regs[REG_PC] = lpc55xx_get_flash_erase_address(target);
+		break;
+	case IAP_CMD_FLASH_PROGRAM:
+		regs[REG_PC] = lpc55xx_get_flash_program_address(target);
+		break;
+	case IAP_CMD_FFR_INIT:
+		regs[REG_PC] = lpc55xx_get_ffr_init_address(target);
+		break;
+	case IAP_CMD_FFR_GET_UUID:
+		regs[REG_PC] = lpc55xx_get_ffr_get_uuid_address(target);
+		break;
+	default:
+		DEBUG_WARN("LPC55xx: bad IAP command\n");
+		return IAP_STATUS_FLASH_INVALID_ARGUMENT;
+	}
+
+	/* Setting a dummy LR does not seem to work as it makes the target
+	 * hard-fault. Instead, set LR to a word known to contain the BKPT
+	 * instruction, so that we can safely halt on IAP function return.
+	 */
+
+	target_mem_write16(target, LPC55xx_CODE_PATCH_ADDRESS, ARM_THUMB_BREAKPOINT);
+	regs[REG_LR] = LPC55xx_CODE_PATCH_ADDRESS | 1; // set the ARM thumb call bit
+
+	/* Write the registers to the target and perform the IAP call */
+
+	target_regs_write(target, regs);
+
+	target_halt_resume(target, false);
+
+	while (!target_halt_poll(target, NULL))
+		continue;
+
+	/* Read back the status code from r0 and return */
+
+	target_regs_read(target, regs);
+
+	return (enum iap_status)regs[0];
+}
+
+struct flash_config {
+	uint32_t flash_block_base;
+	uint32_t flash_total_size;
+	uint32_t flash_block_count;
+	uint32_t flash_page_size;
+	uint32_t flash_sector_size;
+
+	uint32_t reserved0[5];
+	uint32_t sysFreqInMHz;
+	uint32_t reserved1[4];
+};
+
+static void lpc55xx_prepare_flash_config(target_s *target, target_addr_t address)
+{
+	/* The flash_config structure is 60 bytes in size, zero it out as that
+	 * is what the SDK does. For some reason you have to fill in the clock
+	 * speed field ("sysFreqInMHz") before flash_init. Set it to 96MHz (?)
+	 */
+
+	struct flash_config config;
```

Comment:
You can make use of named initialisation syntax here (example assumes the typedef from above):
```c
lpc55xx_flash_config config = {
    .sysFreqInMHz = LPC55xx_IAP_FREQ_IN_MHZ,
};
```
Also, please use lower_snake_case for all the member naming

Reply:
Yeah sorry about that, I think it has this case because I copied the name from the NXP SDK and never noticed it was out of place. Will fix!